cachito/web/api_v1.py

```diff
@@ -184,6 +184,70 @@ def get_request(request_id):
     return flask.jsonify(json)
 
 
+def get_latest_request() -> flask.Response:
+    """
+    Retrieve the latest request for a repo_name/ref.
+
+    The latest request will be the one with the highest id.
+
+    :return: a Flask JSON response
+    :rtype: flask.Response
+    :raise NotFound: if the request is not found
+    """
+    # mypy: Connexion ensures that these cannot be None
+    repo_name = cast(str, flask.request.args.get("repo_name"))
+    ref = cast(str, flask.request.args.get("ref"))
+
+    request = _get_latest_request_by_repo_name_and_ref(repo_name, ref)
+
+    if not request:
+        raise NotFound
+
+    return flask.jsonify(request.to_json())
+
+
+def _get_latest_request_by_repo_name_and_ref(repo_name: str, ref: str) -> Optional[Request]:
+    """
+    Retrieve the latest request for a repo_name/ref.
+
+    The latest request will be the one with the highest id. Initially query the DB
+    for a request with:
+      - matching git ref
+      - repo URL that ends with either repo_name or repo_name.git
+      - the highest request_id
+
+    Failing the initial match on repo_name, fall back to iterating over the result set
+    in-order until a match is found (or not found). This is necessary because we don't
+    know the URL scheme/netloc from the repo_name.
+
+    :param str repo_name: the namespaced repository name
+    :param str ref: the git ref
+    :return: a Request object or None
+    :rtype: Request or None
+    """
+    # The .git ending *may* be present at the end of the Request.repo URL
+    repo_name_with_git = f"{repo_name}.git"
+
+    query = (
+        Request.query.filter(Request.ref == ref)
+        .filter(or_(Request.repo.endswith(repo_name), Request.repo.endswith(repo_name_with_git)))
+        .order_by(desc(Request.id))
+    )
+
+    # Check the first result for a matching repo_name
+    first_request = query.first()
+    if first_request and get_repo_name(first_request.repo) == repo_name:
+        return first_request
+
+    # Fall back to iterating over the full result set, loading it in chunks,
+    # looking for a matching repo_name
+    for request in query.yield_per(10):
+        if get_repo_name(request.repo) == repo_name:
+            return request
```

Comment:
There's clearly something I'm missing here about those requests, so please clarify the following to me:
- we filter requests by a given commit ref
- we then filter those results to limit the set to only those which are related to `repo_name` 
- finally we order the set by highest request ID

So far so good. But why do we need to check  `get_repo_name(first_request.repo) == repo_name` again? Doesn't the SQL filter take care of that already?
Is it because of git protocol in the URI? Or is it because e.g. `"repo": "https://github.com/git/org/eggs"` and `"repo": "https://github.com/org/eggs"` are 2 different repos?

Also, how slow is loading those requests by pages of 10 results? I'm only asking since you're special-handling the first entry and then falling back to iterating over the rest of the set instead of just iterating it right away. What's the performance gain here on this API? Do we plan on running it that often instead only occasionally in a cron job?

Reply:
> Also, how slow is loading those requests by pages of 10 results? I'm only asking since you're special-handling the first entry and then falling back to iterating over the rest of the set instead of just iterating it right away. What's the performance gain here on this API? Do we plan on running it that often instead only occasionally in a cron job?

I expect that the initial query will cover nearly all requests and that the fall-back will just handle the odd edge-case. We'll probably just be running this as-needed and when it comes time to write the script, my plan was to rate-limit the requests anyway to avoid hammering the API for a non-critical maintenance task.